### DISCUSSION TITLE
Build tasks project targeting multiple MSBuild releases

### DISCUSSION THREAD

**Discussion by jozefizso** [dotnet/msbuild#8652] [Q&A] [answered]
Hi, what's the best practise for a project with build tasks which should target multiple MSBuild releases - 15, 16 and 17.

We have a project which compiles build tasks using .NET Framework 4.6 and targets MSBuild 15 only.
We would like to upgrade it to support MSBuild 15, 16 and 17 and target .NET Framework 4.8 and .NET 6.

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net46</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Build.Framework" Version="15.1.1012" />
    <PackageReference Include="Microsoft.Build.Utilities.Core" Version="15.1.1012" />
  </ItemGroup>
</Project>
```

Is is possible to do it in a single project, or should we use shared project for source code and create individual projects for each MSBuild release?

Can this be packed into a single nuget package?

**Comment by rainersigwald** [comment, marked as answer]
Do you need/want to use new features in the newer MSBuild versions? In general, it's easiest to build tasks targeting the lowest MSBuild version you care to support and rely on MSBuild to keep them compatible on newer engines.

**Reply by jozefizso** [reply]
No, the task does not use anything from MSBuild besides logging.

**Reply by rainersigwald** [reply]
Then I'd multitarget it (to add support for the .NET SDK/`dotnet build`) but not change the version of the MSBuild reference.

You can pack into a single NuGet package, but you'll need to have both task assemblies there and to set up conditions on your `UsingTask` elements to select the right copy based on `$(MSBuildRuntimeType)`.
